src/algorithms/libs/beidou_b1c_signal_processing.cc

```diff
@@ -0,0 +1,427 @@
+
+    if (_prn > 0 and _prn < 63)
+        {
+            make_b1cp(_code, _prn);
+        }
+
+    for (signed int i = 0; i < BEIDOU_B1cp_CODE_LENGTH_CHIPS; i++)
+        {
+            _dest[i] = 1.0 - 2.0 * static_cast<float>(_code[i]);
+        }
+
+    delete[] _code;
+}
+
+//! Generates a complex version of the B1c Pilot Primary code
+void beidou_b1cp_code_gen_complex(std::complex<float>* _dest, unsigned int _prn)
+{
+    int32_t* _code = new int32_t[BEIDOU_B1cp_CODE_LENGTH_CHIPS];
+
+    if (_prn > 0 and _prn < 63)
+        {
+            make_b1cp(_code, _prn);
+        }
+
+    for (signed int i = 0; i < BEIDOU_B1cp_CODE_LENGTH_CHIPS; i++)
+        {
+            _dest[i] = std::complex<float>(1.0 - 2.0 * _code[i], 0.0);
+        }
+
+    delete[] _code;
+}
+
+
+//! Generates complex BEIDOU B1c Pilot Primary code for the desired SV ID and sampled to specific sampling frequency
+void beidou_b1cp_code_gen_complex_sampled(std::complex<float>* _dest, unsigned int _prn, signed int _fs)
+{
+    int32_t* _code = new int32_t[BEIDOU_B1cp_CODE_LENGTH_CHIPS];
+    if (_prn > 0 and _prn < 63)
+        {
+            make_b1cp(_code, _prn);
+        }
+
+    signed int _samplesPerCode, _codeValueIndex;
+    float _ts;
+    float _tc;
+    const signed int _codeLength = BEIDOU_B1cp_CODE_LENGTH_CHIPS;
+
+    //--- Find number of samples per spreading code ----------------------------
+    _samplesPerCode = static_cast<int>(static_cast<double>(_fs) / (static_cast<double>(BEIDOU_B1cp_CODE_RATE_HZ) / static_cast<double>(_codeLength)));
+
+    //--- Find time constants --------------------------------------------------
+    _ts = 1.0 / static_cast<float>(_fs);                   // Sampling period in sec
+    _tc = 1.0 / static_cast<float>(BEIDOU_B1cp_CODE_RATE_HZ);  // C/A chip period in sec
+
+    //float aux;
+    for (signed int i = 0; i < _samplesPerCode; i++)
+        {
+            //=== Digitizing =======================================================
+
+            //--- Make index array to read B1c pilot code values -------------------------
+            _codeValueIndex = ceil((_ts * (static_cast<float>(i) + 1)) / _tc) - 1;
+
+            //--- Make the digitized version of the B1c code -----------------------
+            if (i == _samplesPerCode - 1)
+                {
+                    //--- Correct the last index (due to number rounding issues) -----------
+                    _dest[i] = std::complex<float>(1.0 - 2.0 * _code[_codeLength - 1], 0);
+                }
+            else
+                {
+                    _dest[i] = std::complex<float>(1.0 - 2.0 * _code[_codeValueIndex], 0);  //repeat the chip -> upsample
+                }
+        }
+    delete[] _code;
+}
+
+//=========================================SECONDARY CODE GENERATION OF PILOT COMPONENT=========================================
+
+//! Generate a version of the B1c Pilot Secondary code
+void make_b1cpSecondary(int32_t* _dest, int prn)
```

Comment:
can you rename this to make_b1cp_secondary?. Better to keep consistency in the naming. Errors in B2a has been fixed